aiostream/aiter_utils.py

```diff
@@ -46,16 +47,38 @@
 # Magic method shorcuts
 
 
+UNSET = object()
+
+
 def aiter(obj: AsyncIterable[T]) -> AsyncIterator[T]:
     """Access aiter magic method."""
     assert_async_iterable(obj)
     return obj.__aiter__()
 
 
+@overload
 def anext(obj: AsyncIterator[T]) -> Awaitable[T]:
+    pass
+
+
+@overload
+def anext(obj: AsyncIterator[T], default: U) -> Awaitable[T] | U:
+    pass
+
+
+def anext(obj: AsyncIterator[T], default: Any = UNSET) -> Awaitable[T] | Any:
     """Access anext magic method."""
     assert_async_iterator(obj)
-    return obj.__anext__()
+
+    async def anext_default_handling_wrapper():
+        try:
+            return await obj.__anext__()
+        except StopAsyncIteration:
+            if default is not UNSET:
+                return default
+            raise
+
+    return anext_default_handling_wrapper()
```

Comment:
Not sure if you'll agree with this choice, but [Python's own `anext`](https://docs.python.org/3/library/functions.html#anext) (which I understand you can't use in your library because it was only introduced in 3.10) has an optional `default` parameter that, if we could use it, would let us avoid creating an ad-hoc default-returning wrapper around `anext` for the `strict` branch like you did in your suggestion from https://github.com/vxgmichel/aiostream/issues/118#issuecomment-2217381371.

I tried to match the type signature of Python's own `anext` using `@overload`s.

I don't know if the implementation itself is properly typed; Mypy didn't seem to complain, but then, I got a lot of errors from the rest of the codebase which I didn't touch, so maybe I'm not running it correctly...

Reply:
>  has an optional default parameter

Oh I didn't know that, great idea :+1: 

> this probably worsens performance a tiny little bit

Yes `anext` is used a lot in the internal logic, so we should be a bit careful about that.

But you can easily work around it by using a shortcut:
```python
    assert_async_iterator(obj)
    if default is UNSET:
        return obj.__anext__()
    [...]
```

> I got a lot of Mypy errors from the rest of the codebase which I didn't touch

Yes we're sadly not compatible with mypy > 1.6.1:
https://github.com/vxgmichel/aiostream/blob/5e0547afcebfcb556e011df3c185f69794a187cb/.pre-commit-config.yaml#L18-L20

Latest pyright shoud work though.

Reply:
Perfect, thanks! And yes, the shortcut should be much better than what I had; changed it to use that now.

And I see @ Mypy. I'll start a new review thread for the remaining type errors coming from my changes.